### PR TITLE
Add ARKODE unit test for number of mass matrix solves

### PR DESCRIPTION
**Pull request by Steven-Roberts**
Adds a unit test that will allow us to monitor changes to and ensure the correct number of mass matrix solves in ARKODE. I've included comments in each test which explain where the solves arise and if it can be sharpened. The tests cover a range of methods and settings, but let me know if there other combinations worth including. Here is the test output: 

```
Mass Matrix Solve Count Test

Result | Implicit Method               | Explicit Method               | Time-Dep | Deduce Im RHS | Actual | Expected
-------+-------------------------------+-------------------------------+----------+---------------+--------+---------
  Pass | ARKODE_SDIRK_2_1_2            | ARKODE_ERK_NONE               | No       | No            | 2      | 2
  Pass | ARKODE_SDIRK_2_1_2            | ARKODE_ERK_NONE               | No       | Yes           | 2      | 2
  Pass | ARKODE_SDIRK_2_1_2            | ARKODE_ERK_NONE               | Yes      | No            | 2      | 2
  Pass | ARKODE_SDIRK_2_1_2            | ARKODE_ERK_NONE               | Yes      | Yes           | 0      | 0
  Pass | ARKODE_SDIRK_5_3_4            | ARKODE_ERK_NONE               | No       | No            | 1      | 1
  Pass | ARKODE_SDIRK_5_3_4            | ARKODE_ERK_NONE               | No       | Yes           | 1      | 1
  Pass | ARKODE_SDIRK_5_3_4            | ARKODE_ERK_NONE               | Yes      | No            | 5      | 5
  Pass | ARKODE_SDIRK_5_3_4            | ARKODE_ERK_NONE               | Yes      | Yes           | 0      | 0
  Pass | ARKODE_ESDIRK324L2SA_4_2_3    | ARKODE_ERK_NONE               | No       | No            | 2      | 2
  Pass | ARKODE_ESDIRK324L2SA_4_2_3    | ARKODE_ERK_NONE               | No       | Yes           | 2      | 2
  Pass | ARKODE_ESDIRK324L2SA_4_2_3    | ARKODE_ERK_NONE               | Yes      | No            | 4      | 4
  Pass | ARKODE_ESDIRK324L2SA_4_2_3    | ARKODE_ERK_NONE               | Yes      | Yes           | 1      | 1
  Pass | ARKODE_DIRK_NONE              | ARKODE_SHU_OSHER_3_2_3        | No       | No            | 5      | 5
  Pass | ARKODE_DIRK_NONE              | ARKODE_SHU_OSHER_3_2_3        | No       | Yes           | 5      | 5
  Pass | ARKODE_DIRK_NONE              | ARKODE_SHU_OSHER_3_2_3        | Yes      | No            | 3      | 3
  Pass | ARKODE_DIRK_NONE              | ARKODE_SHU_OSHER_3_2_3        | Yes      | Yes           | 3      | 3
  Pass | ARKODE_DIRK_NONE              | ARKODE_BOGACKI_SHAMPINE_4_2_3 | No       | No            | 5      | 5
  Pass | ARKODE_DIRK_NONE              | ARKODE_BOGACKI_SHAMPINE_4_2_3 | No       | Yes           | 5      | 5
  Pass | ARKODE_DIRK_NONE              | ARKODE_BOGACKI_SHAMPINE_4_2_3 | Yes      | No            | 4      | 4
  Pass | ARKODE_DIRK_NONE              | ARKODE_BOGACKI_SHAMPINE_4_2_3 | Yes      | Yes           | 4      | 4
  Pass | ARKODE_ARK2_DIRK_3_1_2        | ARKODE_ARK2_ERK_3_1_2         | No       | No            | 3      | 3
  Pass | ARKODE_ARK2_DIRK_3_1_2        | ARKODE_ARK2_ERK_3_1_2         | No       | Yes           | 3      | 3
  Pass | ARKODE_ARK2_DIRK_3_1_2        | ARKODE_ARK2_ERK_3_1_2         | Yes      | No            | 6      | 6
  Pass | ARKODE_ARK2_DIRK_3_1_2        | ARKODE_ARK2_ERK_3_1_2         | Yes      | Yes           | 4      | 4
  Pass | ARKODE_ARK548L2SA_DIRK_8_4_5  | ARKODE_ARK548L2SA_ERK_8_4_5   | No       | No            | 3      | 3
  Pass | ARKODE_ARK548L2SA_DIRK_8_4_5  | ARKODE_ARK548L2SA_ERK_8_4_5   | No       | Yes           | 3      | 3
  Pass | ARKODE_ARK548L2SA_DIRK_8_4_5  | ARKODE_ARK548L2SA_ERK_8_4_5   | Yes      | No            | 16     | 16
  Pass | ARKODE_ARK548L2SA_DIRK_8_4_5  | ARKODE_ARK548L2SA_ERK_8_4_5   | Yes      | Yes           | 9      | 9

```

### CODE REVIEW

File: test/unit_tests/arkode/C_serial/ark_test_mass.c [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+int main(int argc, char *argv[])
+{
+  int retval = 0;
+
+  printf("Mass Matrix Solve Count Test\n\n");
+  printf("Result | Implicit Method               | Explicit Method               | Time-Dep | Deduce Im RHS | Actual | Expected\n");
+  printf("-------+-------------------------------+-------------------------------+----------+---------------+--------+---------\n");
+  
+  /* SDIRK */
+  /* 1 for primiary and 1 for embedded soltuion. Optimally, could be 0 solves */
```
Comment (drreynolds) [inline test/unit_tests/arkode/C_serial/ark_test_mass.c:206]:
```suggestion
  /* 1 for primary and 1 for embedded solution. Optimally, could be 0 solves */
```

File: test/unit_tests/arkode/C_serial/ark_test_mass.c [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+  retval += solve("ARKODE_SDIRK_2_1_2", "ARKODE_ERK_NONE",
+                  SUNFALSE, SUNFALSE, 2);
+  /* 1 for primiary and 1 for embedded soltuion. Optimally, could be 0 solves */
```
Comment (drreynolds) [inline test/unit_tests/arkode/C_serial/ark_test_mass.c:209]:
```suggestion
  /* 1 for primary and 1 for embedded solution. Optimally, could be 0 solves */
```

File: test/unit_tests/arkode/C_serial/ark_test_mass.c [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+                  SUNFALSE, SUNTRUE, 2);
+  /* One per stage */
+  retval += solve("ARKODE_ESDIRK324L2SA_4_2_3", "ARKODE_ERK_NONE",
+                  SUNTRUE, SUNFALSE, 4);
+  /* Optimal */
+  retval += solve("ARKODE_ESDIRK324L2SA_4_2_3", "ARKODE_ERK_NONE",
+                  SUNTRUE, SUNTRUE, 1);
+
+  /* ERK */
+  /* 1 per stage, 1 for primary solution, and 1 for embedded soltuion.
```
Comment (drreynolds) [inline test/unit_tests/arkode/C_serial/ark_test_mass.c:256]:
```suggestion
  /* 1 per stage, 1 for primary solution, and 1 for embedded solution.
```

File: test/unit_tests/arkode/C_serial/ark_test_mass.c [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+   * Optimally, could be 3 solves */
+  retval += solve("ARKODE_DIRK_NONE", "ARKODE_SHU_OSHER_3_2_3",
+                  SUNFALSE, SUNFALSE, 5);
+  /* 1 per stage, 1 for primary solution, and 1 for embedded soltuion.
```
Comment (drreynolds) [inline test/unit_tests/arkode/C_serial/ark_test_mass.c:260]:
```suggestion
  /* 1 per stage, 1 for primary solution, and 1 for embedded solution.
```

File: test/unit_tests/arkode/C_serial/ark_test_mass.c [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+                  SUNFALSE, SUNTRUE, 5);
+  /* 1 per stage. Optimal */
+  retval += solve("ARKODE_DIRK_NONE", "ARKODE_SHU_OSHER_3_2_3",
+                  SUNTRUE, SUNFALSE, 3);
+  /* 1 per stage. Optimal */
+  retval += solve("ARKODE_DIRK_NONE", "ARKODE_SHU_OSHER_3_2_3",
+                  SUNTRUE, SUNTRUE, 3);
+
+  /* FSAL ERK */
+  /* 1 per stage and 1 for embedded soltuion. 0 needed for primary due to FSAL
```
Comment (drreynolds) [inline test/unit_tests/arkode/C_serial/ark_test_mass.c:272]:
```suggestion
  /* 1 per stage and 1 for embedded solution. 0 needed for primary due to FSAL
```

File: test/unit_tests/arkode/C_serial/ark_test_mass.c [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+   * property. Optimally, could be 4 solves */
+  retval += solve("ARKODE_DIRK_NONE", "ARKODE_BOGACKI_SHAMPINE_4_2_3",
+                  SUNFALSE, SUNFALSE, 5);
+  /* 1 per stage and 1 for embedded soltuion. 0 needed for primary due to FSAL
```
Comment (drreynolds) [inline test/unit_tests/arkode/C_serial/ark_test_mass.c:276]:
```suggestion
  /* 1 per stage and 1 for embedded solution. 0 needed for primary due to FSAL
```

**Review by drreynolds — APPROVED** [review]
Some minor comment typos, but otherwise this looks great.

**Comment by balos1** [comment]
This looks good, but lets hold it until after some of the other PRs are merged.

**Pull request closed without merging**